### DISCUSSION TITLE
"Differential" tracking

### DISCUSSION THREAD

**Discussion by lfarv** [atcollab/at#395] [Ideas]
This is a proposal following the discussions in #317 and #372 concerning the particle coordinates in AT.

Consider a new function very similar to `lattice_pass`/`ringpass`, with an additional  "reference" argument. The signature of this function would be like:
```
xout = f(ring,reference,xin,nturns=1,refpts=None,...) 
```
Its behaviour would be:
- add `reference` to `xin`
- track in parallel `reference` and `(xin+reference)`
- subtract the tracked reference from the tracked particles to get `xout`

So instead of dealing with particles, the function deals with deviations from a reference particle.

The main ideas are:
- a consistent interface: both input and output are expressed as deviations, not coordinates. It would be misleading to have an absolute one  and a relative one.
- the function name (to be defined) and the explicit "reference" argument should avoid any ambiguity about the coordinate system: it makes clear that the function deals with deviations from a reference, not absolute coordinates,
- It treats all 6 coordinates in the same way,
- It works with or without longitudinal motion (4D or 6D),
- it's general since "reference" may be any trajectory, though obviously the main use would be to take the closed orbit.

This function should answer some requests from @swhite2401 and @simoneliuzzo:
when taking the closed orbit as reference, all 6 coordinates are oscillating around zero, thus eliminating all offsets and drifts. I understand that this may be interesting for plotting or for frequency analysis (though it does not strictly eliminate the DC components).

Notes:
keeping the reference explicit rather that computing internally the closed orbit has the following advantages:
- make the function more general (not really important, there is very small interest of using something else that closed orbit or zero)
- avoid recomputing the closed orbit which is often already available,
- leave the closed orbit parameters open: `dp`, `dct` and other parameters,
- avoid confusion with `lattice_pass` or `ringpass`

**Comment by swhite2401** [comment]
@lfarv thanks for looking into this. 
I did think about it as well and agree with the method.

I have a stupid question though: why not simply adding an optional argument to `lattice_pass`?

I think users already are confused with `linepass, ringpass, lattice_pass, atpass` etc... and reducing the number of functions for tracking to one in python was a huge improvement. I would therefore be very much in favor of keeping it like this.

You may think of a very simple interface : `lattice_pass(ring, rin, refpts=None, refpart=None,...)`
`refpart` are the initial 6D coordinates of the reference particle, when `None` nothing is done, otherwise subtract the trajectory of the reference particle from `rout`.

You may even think of defining an 'auto' mode that takes automatically the closed orbit as the reference particle.
